### ISSUE TITLE
tmux CPU usage high with poweline

### ISSUE THREAD

**Issue by bedge** [tmux/tmux#361]
A few months ago tmux made a jump in CPU utilization. At the time I didn't have the BW to spend any time on this so I ignored it. 
Now that it's been happening for some time, I have no idea what may have changed at that time to cause this.

![screen shot 2016-04-06 at 12 00 36 pm](https://cloud.githubusercontent.com/assets/499317/14333883/b22ebfda-fc05-11e5-8061-ecdf8d3846db.png)

It does only happen when I use tmux in conjunction with powerline. I originally filed a powerline bug: https://github.com/powerline/powerline/issues/1564 and they referred me to tmux.

If I run with -v, I see a lot of these in the log:

> 1459979040.179888 got 105 from client 25
> 1459979040.179891 got 105 from client 25
> 1459979040.179894 got 105 from client 25
> 1459979040.179898 got 105 from client 25
> 1459979040.179901 got 105 from client 25

Taking a stab in the dark and grepping for powerline in the log, I see a lot of these:

> 1459979135.823895 format 'env "$POWERLINE_COMMAND" $POWERLINE_COMMAND_ARGS tmux right -R pane_id="`tmux display -p "#D"`" --width=`tmux display -p "#{client_width}"` -R width_adjust=`tmux show-options -g status-left-length | cut -d" " -f 2`' -> 'env "$POWERLINE_COMMAND" $POWERLINE_COMMAND_ARGS tmux right -R pane_id="`tmux display -p "%4"`" --width=`tmux display -p "277"` -R width_adjust=`tmux show-options -g status-left-length | cut -d" " -f 2`'
> 1459979135.823912 format '#(env "$POWERLINE_COMMAND" $POWERLINE_COMMAND_ARGS tmux right -R pane_id="`tmux display -p "#D"`" --width=`tmux display -p "#{client_width}"` -R width_adjust=`tmux show-options -g status-left-length | cut -d" " -f 2`)' -> '#[fg=colour233,bg=default,nobold,noitalics,nounderscore]\302\240\356\202\262#[fg=colour247,bg=colour233,nobold,noitalics,nounderscore]\302\240\342\207\221\302\240\302\2407d\302\2406h\302\24055m#[fg=colour241,bg=colour233,nobold,noitalics,nounderscore]\302\240\356\202\263#[fg=colour2,bg=colour233,nobold,noitalics,nounderscore]\302\2402.8\302\240#[fg=colour2,bg=colour233,nobold,noitalics,nounderscore]3.2\302\240#[fg=colour2,bg=colour233,nobold,noitalics,nounderscore]3.3#[fg=colour236,bg=colour233,nobold,noitalics,nounderscore]\302\240\356\202\262#[fg=colour247,bg=colour236,nobold,noitalics,nounderscore]\302\2402016-04-06#[fg=colour241,bg=colour236,nobold,noitalics,nounderscore]\302\240\356\202\263#[fg=colour252,bg=colour236,bold,noitalics,nounderscore]\302\240\342\214\232\302\24014:45#[fg=colour252,bg=colour236,nobold,noitalics,nounderscore]\302\240\356\202\262#[fg=colour16,bg=colour252,bold,noitalics,nounderscore]\302\240\356\202\242\302\240tmac.local\302\240'

Here's my comment-stripped .tmux.conf:

> set-option -g prefix `
> unbind-key C-b
> bind-key C-a last-window
> bind-key ` last-window
> bind-key a send-prefix
> set -g set-remain-on-exit on
> bind C-r respawn-pane
> bind R respawn-window
> bind-key F11 set-option -g prefix C-a
> bind-key F12 set-option -g prefix `
> set -g base-index 1
> setw -g pane-base-index 1
> set -g xterm-keys on
> set -g mouse-utf8 on
> set -g mouse on
> bind m \
>     set -g mouse on \;\
>     display 'Mouse: ON'
> bind M \
>     set -g mouse off \;\
>     display 'Mouse: OFF'
> setw -g mode-keys vi
> setw -g monitor-activity on
> set -g default-command "reattach-to-user-namespace -l ${SHELL}"
> set -g default-terminal "screen-256color"
> setw -g aggressive-resize on
> 
> TMROOT=$GIT_ROOT/server/content/sdk
> SESSION=b-ct
> HIST_ID=$SESSION.main
> new -n $SESSION -s $SESSION
> neww -n content		"cd $TMROOT/;					HIST_ID=$SESSION.content	tmux.sh -l $SHELL"
> neww -n parser-v4	"cd $TMROOT/content-parser-v4;	HIST_ID=$SESSION.parser-v4 	tmux.sh -l $SHELL"
> neww -n cm			"cd $TMROOT/content-manager; 	HIST_ID=$SESSION.cm			tmux.sh -l $SHELL"
> neww -n cb			"cd $TMROOT/content-builder; 	HIST_ID=$SESSION.cb			tmux.sh -l $SHELL"
>
> set -g @plugin 'tmux-plugins/tpm'
> set -g @plugin 'tmux-plugins/tmux-sensible'
> set -g @plugin 'tmux-plugins/tmux-sensible'
> set -g @plugin 'tmux-plugins/tmux-pain-control'
> set -g @plugin 'tmux-plugins/tmux-sessionist'
> set -g @plugin 'tmux-plugins/tmux-copycat'
> set -g @plugin 'tmux-plugins/tmux-yank'
> set -g @plugin 'tmux-plugins/tmux-logging'
> run '~/.tmux/plugins/tpm/tpm'

The above "tmux.sh" is a symlink:
> /Users/bedge/bin/tmux.sh -> /usr/local/bin/reattach-to-user-namespace
ie: Just a way to get the osx copy/paste stuff working again.

The end of my .zshrc has this:

> USER_PYTHON_LIB=/Users/bedge/Library/Python/2.7/lib/python/site-packages
> if [[ "$TMUX" ]] && [[ -f ${USER_PYTHON_LIB}/powerline/bindings/tmux/powerline.conf ]]; then
>     tmux source "${USER_PYTHON_LIB}/powerline/bindings/tmux/powerline.conf"
> fi

A post I found detailed this as the "right way" to configure tmux & powerline in osx.
Although, this is not what I had originally it did get it working again after I rolled everything to current versions before posting this.
This in .tmux.conf doesn't appear to work anymore:

> source /Users/bedge/Library/Python/2.7/lib/python/site-packages/powerline/bindings/tmux/powerline.conf

Here's the powerline & version info:

> 
> %C> pip show powerline-status
> ---
> Metadata-Version: 2.0
> Name: powerline-status
> Version: 2.3
> Summary: The ultimate statusline/prompt utility.
> Home-page: https://github.com/powerline/powerline
> Author: Kim Silkebaekken
> Author-email: kim.silkebaekken+vim@gmail.com
> Installer: pip

> 
> 
> %C> brew info tmux
> tmux: stable 2.1 (bottled), HEAD
> Terminal multiplexer
> https://tmux.github.io/
> /usr/local/Cellar/tmux/1.9a (16 files, 599.2K)
>   Poured from bottle
> /usr/local/Cellar/tmux/2.0 (16 files, 620.6K)
>   Poured from bottle
> /usr/local/Cellar/tmux/2.1 (15 files, 648.8K) *
>   Poured from bottle
> From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/tmux.rb

I expect it's the result of a mixture of old & new config & tooling. It's hard to know where to draw the line in what to include as tmux tends to infiltrate some pretty convoluted setups.
Anyway, let me know if I've missed anything vital, or if just need to RTFM.

**Comment by nicm** [comment]
Please reproduce without powerline, I suggest you copy what it puts in
status-left and status-right and so on and see if using them directly shows
the same problem.

**Comment by ThomasAdam** [comment]
I suspect this is related to the way tmux handles jobs.

**Comment by bedge** [comment]
I've tried various ways to to reproduce exactly what powerline if configuring tmux to do , but the config is fairly convoluted.

However, one thing I have noticed is that it appears that with powerline activated, tmux seems to be stuck in an initialization loop, or at least that's what it appears from the ps output, or, maybe it's a repetitive powerline update queries, I don't know.

Here's what ps shows with no powerline active:

```
0 [5] 12:37:53 bedge@tmac ~
%> ps ax | grep tmux
79306   ??  Ss     0:15.87 tmux attach
79304 s000  S+     0:00.04 tmux attach
88540 s005  R+     0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn tmux

```

Then, after I manually start the powerline config using:
`tmux source /usr/local/lib/python2.7/site-packages/powerline/bindings/tmux/powerline.conf`

Then the ps output shows this:

```
%> ps ax | grep tmux
79306   ??  Us     0:24.38 tmux attach
88781   ??  S      0:00.00 sh -c env \$POWERLINE_COMMAND \$POWERLINE_COMMAND_ARGS tmux left --width=`tmux display -p '277'` -R width_adjust=`tmux show-options -g status-right-length | cut -d' ' -f2` -R pane_id="`tmux display -p '%4'`"
88784   ??  S      0:00.00 sh -c env "$POWERLINE_COMMAND" $POWERLINE_COMMAND_ARGS tmux right -R pane_id="`tmux display -p "%4"`" --width=`tmux display -p "277"` -R width_adjust=`tmux show-options -g status-left-length | cut -d" " -f 2`
88787   ??  S      0:00.00 sh -c env \$POWERLINE_COMMAND \$POWERLINE_COMMAND_ARGS tmux left --width=`tmux display -p '277'` -R width_adjust=`tmux show-options -g status-right-length | cut -d' ' -f2` -R pane_id="`tmux display -p '%4'`"
88788   ??  R      0:00.01 tmux show-options -g status-right-length
88792   ??  S      0:00.00 sh -c env "$POWERLINE_COMMAND" $POWERLINE_COMMAND_ARGS tmux right -R pane_id="`tmux display -p "%4"`" --width=`tmux display -p "277"` -R width_adjust=`tmux show-options -g status-left-length | cut -d" " -f 2`
88793   ??  S      0:00.00 tmux display -p 277
79304 s000  S+     0:00.04 tmux attach
88799 s005  R+     0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn tmux

```
It never stabilizes after this. These additional tmux threads are always active with different PIDs each time.

Is this the powerline polling, or is something terminating abnormally in the init and consequently stuck in this loop?

**Comment by nicm** [comment]
tmux will run commands in the various status options as often as it is
asked to up to a maximum of once a second (for each command).

**Comment by bedge** [comment]
Is there a simple procedure to figure out what the tmux native equivalent config would be to replicate the powerline config?
ie: to try accomplish the first suggestion, which was to replicate the problem using tmux only.

**Comment by ntnn** [comment]
You'll have to build it yourself, if it's too messy put it in
a shellscripts. Take care not to put too much in one `#()` or the whole
expression won't be displayed - or it times out.

E.g., this displays the current mpd track if it plays or is paused, the
temperature if it exists and the loadaverage.

    set -g status-right " #(mpc current) "
    set -ag status-right " #({rev | cut -c 4- | rev } < /sys/devices/virtual/thermal/thermal_zone0/temp) "
    set -ag status-right " #(awk '{print $1, $2, $3}' < /proc/loadavg) "
    set -ag status-right " #H %T %F"

You'll have to see how and what each of the modules you enabled in
powerline does and either recreate it or use the tool without powerline.

E.g. assume the weather module is a wrapper for a python script you can
install via pip - depending on the output you can either use it
`#(weather-plugin)` or write a script to parse the output and use that
`#(weather-plugin-wrapper)`.

So _simple_ depends on which modules you activated.

**Comment by nicm** [comment]
I don't know, you will just have to play around and see if you can find out
what powerline is doing that causes this. I would start by running it
normally then you can inspect status-left etc and either remove them from
the running server to see if the problem goes away or copy them to a new
tmux server you are running at the same time and see if it starts
misbehaving (for example, start one with: tmux -Ltest -f/dev/null new).

I am limited in what I can tell you because I don't know anything about
powerline and I do not have the time to try it at the moment. I can only
really help you with tmux problems and you will need to narrow it down.
There was a change a while ago to move jobs into formats but so far nobody
else has reported any issues with it. You could try to git bisect I
suppose, if you know how, and see if it started with a specific commit.

**Comment by bedge** [comment]
Just to followup, I ended up ditching powerline. I replicated an almost identical setup in tmux config directly and there's no CPU issue. Maybe related to powerline needing to fire up a python interpreter for each refresh? Can't say for sure.
I was going to instrument the powerline exec calls to see where they were spending all their time, but I kinda lost interest when the tmux only setup worked fine.
Thanks @nicm for your guidance. Your comments were very helpful.
